File: src/main/java/visitor/user/Teacher.java
package visitor.user;

import java.math.BigDecimal;
import visitor.visitor.Visitor;

/** @author Elias (siran0611@gmail.com) */
public class Teacher extends User {

  public Teacher(String name, String identity, String clazz) {
    super(name, identity, clazz);
  }

  @Override
  public void accept(Visitor visitor) {
    visitor.visit(this);
  }

  public double entranceRatio() {
    return BigDecimal.valueOf(Math.random() * 100)
        .setScale(2, BigDecimal.ROUND_HALF_UP)
        .doubleValue();
  }
}

File: src/rust/visitor/user/teacher.rs
use crate::rust::visitor::user::{User, UserEnum};
use crate::rust::visitor::visitor::Visitor;

#[derive(Debug, Clone)]
pub struct Teacher {
    pub name: &'static str,
    identity: &'static str,
    class: &'static str,
}

impl User for Teacher {
    fn accept(&self, visitor: &impl Visitor) {
        let teacher = UserEnum::Teacher(self.clone());
        visitor.visit(teacher);
    }
}

impl Teacher {
    pub fn new(name: &'static str, identity: &'static str, class: &'static str) -> Self {
        Self { name, identity, class }
    }

    pub fn entrance_ratio(&self) -> f64 {
        0.8
    }
}